controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,202 @@
+		r.Log.Error(err, "device class", storageAutoScaler.Spec.DeviceClass)
+		return reconcile.Result{}, err
+	}
+
+	// check if the highest osd usage is greater than the threshold
+	if !((usage.(float64) * 100) >= float64(storageAutoScaler.Spec.StorageScalingThresholdPercent)) {
+		return reconcile.Result{}, nil
+	}
+	// scaling is required
+	r.Log.Info("scaling is required", "device class", storageAutoScaler.Spec.DeviceClass)
+
+	// calculate the expectedStorageCapacity
+	var startOsdSize, expectedOsdSize resource.Quantity
+	var startOsdCount, expectedOsdCount, deviceSetCount int
+
+	// assuming heterogeneous distribution of osd across device sets
+	// so osd size and count is same for all device sets
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if (deviceSet.DeviceClass == "" && deviceClass == "ssd") || deviceSet.DeviceClass == deviceClass {
+			startOsdSize = deviceSet.DataPVCTemplate.Spec.Resources.Requests["storage"]
+			startOsdCount = deviceSet.Count
+			deviceSetCount++
+		}
+	}
+
+	startStorageCapacity := *resource.NewQuantity(int64(deviceSetCount*startOsdCount)*startOsdSize.Value(), startOsdSize.Format)
+
+	if (resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)).Cmp(storageAutoScaler.Spec.MaxOsdSize) <= 0 {
+		// vertical scaling
+		// double the osd size
+		expectedOsdSize = *resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)
+	} else {
+		// horizontal scaling
+		expectedOsdCount = startOsdCount + 1
+	}
+
+	expectedStorageCapacity := *resource.NewQuantity(int64(deviceSetCount*expectedOsdCount)*expectedOsdSize.Value(), startOsdSize.Format)
+
+	if expectedStorageCapacity.Cmp(storageAutoScaler.Spec.StorageCapacityLimit) > 0 {
+		r.Log.Info("storage capacity limit reached")
+		// update the status
+		bool := true
+		storageAutoScaler.Status.StorageCapacityLimitReached = &bool
+		err := r.updateStatus(ctx, storageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// update the status
+	storageAutoScaler.Status.Phase = "InProgress"
+	storageAutoScaler.Status.LastExpansion = &ocsv1.LastExpansionStatus{
+		StartOsdCount:           uint16(startOsdCount),
+		ExpectedOsdCount:        uint16(expectedOsdCount),
+		StartOsdSize:            startOsdSize,
+		ExpectedOsdSize:         expectedOsdSize,
+		StartStorageCapacity:    startStorageCapacity,
+		ExpectedStorageCapacity: expectedStorageCapacity,
+		StartTime:               metav1.Now(),
+		CompletionTime:          metav1.Time{},
+	}
+	err = r.updateStatus(ctx, storageAutoScaler)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// patch the storagecluster with the expected osd size and count
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if (deviceSet.DeviceClass == "" && deviceClass == "ssd") || deviceSet.DeviceClass == deviceClass {
```

Comment:
This exact condition is earlier in this method, how about factoring to a helper mehtod?

Reply:
it a simpler check lets keep into the flow

Reply:
Code duplication can easily lead to long term bugs, and code is also harder to read in this case. A helper method can be named to make it more readable. For example:
```go
if deviceClassMatchesDeviceSet(deviceSet, deviceClass) {
...
}

func deviceClassMatchesDeviceSet(deviceSet StorageDeviceSet, deviceClass string) bool {
  return (deviceSet.DeviceClass == "" && deviceClass == "ssd") || deviceSet.DeviceClass == deviceClass
}
```
And you can also easily unit test the helper method.